test/memcached/cache_impl_test.go

```diff
@@ -0,0 +1,526 @@
+	client.EXPECT().GetMulti([]string{"domain_key8_value8_1234"}).Return(
+		getMultiResult(map[string]int{"domain_key8_value8_1234": 19}), nil,
+	)
+	client.EXPECT().Increment("domain_key8_value8_1234", uint64(3)).Return(uint64(22), nil)
+
+	request = common.NewRateLimitRequest("domain", [][][2]string{{{"key8", "value8"}}}, 3)
+	limits = []*config.RateLimit{config.NewRateLimit(20, pb.RateLimitResponse_RateLimit_SECOND, "key8_value8", statsStore)}
+
+	assert.Equal(
+		[]*pb.RateLimitResponse_DescriptorStatus{{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0, DurationUntilReset: utils.CalculateReset(limits[0].Limit, timeSource)}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(3), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(2), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(1), limits[0].Stats.NearLimit.Value())
+
+	// Some of it in all three places
+	timeSource.EXPECT().UnixNow().Return(int64(1234)).MaxTimes(3)
+	client.EXPECT().GetMulti([]string{"domain_key9_value9_1234"}).Return(
+		getMultiResult(map[string]int{"domain_key9_value9_1234": 15}), nil,
+	)
+	client.EXPECT().Increment("domain_key9_value9_1234", uint64(7)).Return(uint64(22), nil)
+
+	request = common.NewRateLimitRequest("domain", [][][2]string{{{"key9", "value9"}}}, 7)
+	limits = []*config.RateLimit{config.NewRateLimit(20, pb.RateLimitResponse_RateLimit_SECOND, "key9_value9", statsStore)}
+
+	assert.Equal(
+		[]*pb.RateLimitResponse_DescriptorStatus{{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0, DurationUntilReset: utils.CalculateReset(limits[0].Limit, timeSource)}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(7), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(2), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(4), limits[0].Stats.NearLimit.Value())
+
+	// all of it over limit
+	timeSource.EXPECT().UnixNow().Return(int64(1234)).MaxTimes(3)
+	client.EXPECT().GetMulti([]string{"domain_key10_value10_1234"}).Return(
+		getMultiResult(map[string]int{"domain_key10_value10_1234": 27}), nil,
+	)
+	client.EXPECT().Increment("domain_key10_value10_1234", uint64(3)).Return(uint64(30), nil)
+
+	request = common.NewRateLimitRequest("domain", [][][2]string{{{"key10", "value10"}}}, 3)
+	limits = []*config.RateLimit{config.NewRateLimit(10, pb.RateLimitResponse_RateLimit_SECOND, "key10_value10", statsStore)}
+
+	assert.Equal(
+		[]*pb.RateLimitResponse_DescriptorStatus{{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0, DurationUntilReset: utils.CalculateReset(limits[0].Limit, timeSource)}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(3), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(3), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(0), limits[0].Stats.NearLimit.Value())
+
+	cache.Flush()
+}
+
+func TestMemcacheWithJitter(t *testing.T) {
+	assert := assert.New(t)
```

Comment:
every test sets up the same amount of routine code, this could be extracted to a sort of SetUp() function, to reduce the size of the test body and leave the essential part